src/uki.rs

```diff
@@ -0,0 +1,183 @@
+    ))
+}
+
+fn dequote(value: &str) -> Option<String> {
+    // https://pubs.opengroup.org/onlinepubs/009604499/utilities/xcu_chap02.html
+    let mut result = String::new();
+    let mut iter = value.trim().chars();
+
+    // os-release spec says we don't have to support concatenation of independently-quoted
+    // substrings, but honestly, it's easier if we do...
+    while let Some(c) = iter.next() {
+        match c {
+            '"' => loop {
+                result.push(match iter.next()? {
+                    // Strictly speaking, we should only handle \" \$ \` and \\...
+                    '\\' => iter.next()?,
+                    '"' => break,
+                    other => other,
+                });
+            },
+
+            '\'' => loop {
+                result.push(match iter.next()? {
+                    '\'' => break,
+                    other => other,
+                });
+            },
+
+            // Per POSIX we should handle '\\' sequences here, but os-release spec says we'll only
+            // encounter A-Za-z0-9 outside of quotes, so let's not bother with that for now...
+            other => result.push(other),
+        }
+    }
+
+    Some(result)
+}
+
+// Gets the value that matches the first key that's present and successfully dequotes.  If none are
+// present and/or dequoting fails, fall back on the default.
+fn get_value(map: &HashMap<&str, &str>, keys: &[&str], default: &str) -> String {
+    for key in keys {
+        if let Some(value) = map.get(key).and_then(|v| dequote(v)) {
+            return value;
+        }
+    }
+
+    String::from(default)
+}
+
+/// Gets an appropriate label for display in the boot menu for the given UKI image, according to
+/// the "Type #2 EFI Unified Kernel Images" section in the Boot Loader Specification.  This will be
+/// based on the "PRETTY_NAME" and "VERSION_ID" fields found in the os-release file (falling back
+/// to "ID" and/or "VERSION" if they are not present).
+///
+/// For more information, see:
+///  - https://uapi-group.org/specifications/specs/boot_loader_specification/
+///  - https://www.freedesktop.org/software/systemd/man/latest/os-release.html
+///
+/// # Arguments
+///
+///  * `image`: the complete UKI image as a byte slice
+///
+/// # Return value
+///
+/// If we could successfully parse the provided UKI as a Portable Executable file and find an
+/// ".osrel" section in it, return a string to use as the boootloader entry.  If we were unable to
+/// find any meaningful content in the os-release information this will be "Unknown 0".
+///
+/// If we couldn't parse the PE file or couldn't find an ".osrel" section then an error will be
+/// returned.
+pub fn get_boot_label(image: &[u8]) -> Result<String, UkiError> {
+    let os_release = get_os_release(image)?;
+    let name = get_value(&os_release, &["PRETTY_NAME", "ID"], "Unknown");
+    let version = get_value(&os_release, &["VERSION_ID", "VERSION"], "0");
+    Ok(format!("{name} {version}"))
```

Comment:
Obviously this is a degenerate case but it seems clearer to omit the version if it's empty than to make it 0

Reply:
Agreed, although it makes the thing with the space slightly more annoying.  Maybe I'll just leave the space in :)